File: java/src/main/java/io/github/emindeniz99/applepurchase/receipt/AppReceipt.java
package io.github.emindeniz99.applepurchase.receipt;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * A verified legacy app receipt (PKCS#7 payload). Only receipts returned by
 * {@link ReceiptVerifier} should be trusted — this class carries no proof by
 * itself. {@code null} fields were absent from the receipt.
 */
public final class AppReceipt {

    private final String receiptType;
    private final String bundleId;
    private final byte[] bundleIdBytes;
    private final String appVersion;
    private final byte[] opaqueValue;
    private final byte[] sha1Hash;
    private final Instant creationDate;
    private final Instant originalPurchaseDate;
    private final String originalAppVersion;
    private final Instant expirationDate;
    private final List<InAppPurchase> inAppPurchases;
    private final Map<Integer, List<byte[]>> unknownAttributes;

    AppReceipt(String receiptType, String bundleId, byte[] bundleIdBytes, String appVersion,
               byte[] opaqueValue, byte[] sha1Hash, Instant creationDate,
               Instant originalPurchaseDate, String originalAppVersion,
               Instant expirationDate, List<InAppPurchase> inAppPurchases,
               Map<Integer, List<byte[]>> unknownAttributes) {
        this.receiptType = receiptType;
        this.originalPurchaseDate = originalPurchaseDate;
        this.bundleId = bundleId;
        this.bundleIdBytes = bundleIdBytes;
        this.appVersion = appVersion;
        this.opaqueValue = opaqueValue;
        this.sha1Hash = sha1Hash;
        this.creationDate = creationDate;
        this.originalAppVersion = originalAppVersion;
        this.expirationDate = expirationDate;
        this.inAppPurchases = Collections.unmodifiableList(inAppPurchases);
        this.unknownAttributes = Collections.unmodifiableMap(unknownAttributes);
    }

    /** Attribute 0, e.g. "Production" / "ProductionSandbox" (undocumented). */
    public String receiptType() {
        return receiptType;
    }

    /** Attribute 18 (undocumented; community-established). */
    public Instant originalPurchaseDate() {
        return originalPurchaseDate;
    }

    public String bundleId() {
        return bundleId;
    }

    /** Raw DER bytes of attribute 2 — input to the device-hash check. */
    public byte[] bundleIdBytes() {
        return bundleIdBytes == null ? null : bundleIdBytes.clone();
    }

    public String appVersion() {
        return appVersion;
    }

    /** Attribute 4 — device-specific opaque value used in the hash binding. */
    public byte[] opaqueValue() {
        return opaqueValue == null ? null : opaqueValue.clone();
    }

    /** Attribute 5 — SHA-1 of (device GUID ‖ opaque value ‖ bundle id bytes). */
    public byte[] sha1Hash() {
        return sha1Hash == null ? null : sha1Hash.clone();
    }

    /** Attribute 12 — when Apple signed this receipt. */
    public Instant creationDate() {
        return creationDate;
    }

    /** Attribute 19 — version the user originally purchased. */
    public String originalAppVersion() {
        return originalAppVersion;
    }

    /** Attribute 21 — only present in receipts with an expiry (e.g. VPP). */
    public Instant expirationDate() {
        return expirationDate;
    }

    public List<InAppPurchase> inAppPurchases() {
        return inAppPurchases;
    }

    /**
     * Raw values of attribute types this library does not model, keyed by
     * type — forward compatibility for fields Apple may add (PLAN D10).
     * Values are the raw octet-string contents, verified but undecoded.
     */
    public Map<Integer, List<byte[]>> unknownAttributes() {
        return unknownAttributes;
    }
}

File: dotnet/src/ApplePurchaseReceiptVerifier/Receipt/AppReceipt.cs
using System;
using System.Collections.Generic;
using ApplePurchaseReceiptVerifier.Internal;

namespace ApplePurchaseReceiptVerifier.Receipt
{
    /// <summary>
    /// A verified legacy app receipt (the PKCS#7 payload). Only receipts
    /// returned by <see cref="ReceiptVerifier"/> should be trusted — this class
    /// carries no proof by itself. <see langword="null"/> fields were absent
    /// from the receipt.
    /// </summary>
    public sealed class AppReceipt
    {
        private readonly byte[]? _bundleIdBytes;
        private readonly byte[]? _opaqueValue;
        private readonly byte[]? _sha1Hash;

        internal AppReceipt(
            string? receiptType,
            string? bundleId,
            byte[]? bundleIdBytes,
            string? appVersion,
            byte[]? opaqueValue,
            byte[]? sha1Hash,
            DateTimeOffset? creationDate,
            DateTimeOffset? originalPurchaseDate,
            string? originalAppVersion,
            DateTimeOffset? expirationDate,
            IReadOnlyList<InAppPurchase> inAppPurchases,
            IReadOnlyDictionary<int, IReadOnlyList<byte[]>> unknownAttributes)
        {
            ReceiptType = receiptType;
            BundleId = bundleId;
            _bundleIdBytes = bundleIdBytes;
            AppVersion = appVersion;
            _opaqueValue = opaqueValue;
            _sha1Hash = sha1Hash;
            CreationDate = creationDate;
            OriginalPurchaseDate = originalPurchaseDate;
            OriginalAppVersion = originalAppVersion;
            ExpirationDate = expirationDate;
            InAppPurchases = inAppPurchases;
            UnknownAttributes = unknownAttributes;
        }

        /// <summary>Attribute 0, e.g. "Production" / "ProductionSandbox" (undocumented).</summary>
        public string? ReceiptType { get; }

        /// <summary>Attribute 2.</summary>
        public string? BundleId { get; }

        /// <summary>
        /// Raw DER bytes of attribute 2 — the third input to the device-hash
        /// check. A fresh copy per call, so a caller cannot mutate the receipt.
        /// </summary>
        public byte[]? BundleIdBytes => ByteOps.Copy(_bundleIdBytes);

        /// <summary>Attribute 3.</summary>
        public string? AppVersion { get; }

        /// <summary>Attribute 4 — the device-specific opaque value. A fresh copy per call.</summary>
        public byte[]? OpaqueValue => ByteOps.Copy(_opaqueValue);

        /// <summary>
        /// Attribute 5 — SHA-1 of (device GUID ‖ opaque value ‖ bundle id
        /// bytes). A fresh copy per call.
        /// </summary>
        public byte[]? Sha1Hash => ByteOps.Copy(_sha1Hash);

        /// <summary>Attribute 12 — when Apple signed this receipt.</summary>
        public DateTimeOffset? CreationDate { get; }

        /// <summary>Attribute 18 (undocumented; community-established).</summary>
        public DateTimeOffset? OriginalPurchaseDate { get; }

        /// <summary>Attribute 19 — the version the user originally purchased.</summary>
        public string? OriginalAppVersion { get; }

        /// <summary>Attribute 21 — only present in receipts with an expiry (e.g. VPP).</summary>
        public DateTimeOffset? ExpirationDate { get; }

        /// <summary>Attribute 17, repeated.</summary>
        public IReadOnlyList<InAppPurchase> InAppPurchases { get; }

        /// <summary>
        /// Raw values of attribute types this library does not model, keyed by
        /// type — forward compatibility for fields Apple may add (PLAN D10).
        /// The arrays are this library's own copies of verified receipt bytes;
        /// treat them as read-only.
        /// </summary>
        public IReadOnlyDictionary<int, IReadOnlyList<byte[]>> UnknownAttributes { get; }

        internal byte[]? BundleIdBytesInternal => _bundleIdBytes;

        internal byte[]? OpaqueValueInternal => _opaqueValue;

        internal byte[]? Sha1HashInternal => _sha1Hash;
    }
}
